// digioptions-contracts.js/js/constants.js
// vim: sts=2:ts=2:sw=2
(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD
    define([
      'factsigner'
    ], factory);
  } else if (typeof module !== 'undefined' && module.exports) {
    // CommonJS (node and other environments that support module.exports)
    module.exports = factory(
      require('factsigner')
    );
  }else {
    // Global (browser)
    root.digioptionsContractsConstants = factory(
      root.factsigner
    );
  }
}(this, function (factsigner) {

  return {
    userState: {
      NONE: 0,
      EXISTS: 1,
      PAYED_OUT: 2
    },
    contractType: {
      UNKNOWN: 0,
      DIGIOPTIONSMARKETS: 1,
      DIGIOPTIONSMARKETLISTER: 2
    },
    baseUnitExpDefault: 18,
    // TODO remove
    payoutPerNanoOptionExp: 9, // the smallest buyable fractional option ('nanoOption') results in a potential payout of 10**9 wei
    // TODO remove
    optionPayout: 1000000000, // TODO remove payout per successful option (in wei)

    /* all intervals that may be used by digiOptions */
    marketIntervalsAll: [
      // TODO add all
      factsigner.constants.marketInterval.DAILY,
      factsigner.constants.marketInterval.WEEKLY,
      factsigner.constants.marketInterval.MONTHLY,
      factsigner.constants.marketInterval.YEARLY
    ]
  };
}));
